theverge.com: Subscription banner doesn’t always appear, causing the page to become unusable.

mattwoodrow opened the pull request:
#### def045dbb011ef2b602499998b7de710632d9ca6
<pre>
theverge.com: Subscription banner doesn’t always appear, causing the page to become unusable.
<a href="https://bugs.webkit.org/show_bug.cgi?id=301317">https://bugs.webkit.org/show_bug.cgi?id=301317</a>
<a href="https://rdar.apple.com/157773745">rdar://157773745</a>

Reviewed by Simon Fraser.

This is a race condition, where the UI process sends an async scroll update, but
the page has mutated and is no longer scrollable when it arrives.
The next update from the UI process attempts to fix this, but since it&apos;s a
syncPosition update, it doesn&apos;t clear the existing (and invalid) scroll
position.

syncPosition GraphicsLayer updates should mutate the layer without marking it as
changed (to prevent it being sent back to the UI-process from where it
originated).  They also had the side effect of not overriding existing
approximatePosition updates, which seems to be unexpected. This makes the
behavior match setPosition, where it clears any prior approximate position
updates.

* Source/WebCore/platform/graphics/GraphicsLayer.h:
(WebCore::GraphicsLayer::syncPosition):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::syncPosition):

Canonical link: <a href="https://commits.webkit.org/302159@main">https://commits.webkit.org/302159@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0c6d1a1fe699b82212c9d8ac80cbaedd47056c57

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/128168 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/451 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/39000 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/135541 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/79656 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/d57d9bd5-1deb-4267-94a3-e6246d90f451) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/130040 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/375 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/323 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/97564 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/65458 "") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/ba3f5c67-2883-49f6-a04e-58896efa82f3) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/131116 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/233 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/114816 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/78133 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/b3525873-a14c-4f04-a912-4346785523f1) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/223 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/32923 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/78841 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/108600 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/33408 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/138020 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/303 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/285 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/106093 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/335 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/111158 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/105852 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26988 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/231 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/29712 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/52540 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/349 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/62381 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/256 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/329 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/316 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

smfr reviewed (approved):
I suppose it's hard to write a test.

mattwoodrow committed:
theverge.com: Subscription banner doesn’t always appear, causing the page to become unusable.
https://bugs.webkit.org/show_bug.cgi?id=301317
rdar://157773745

Reviewed by Simon Fraser.

This is a race condition, where the UI process sends an async scroll update, but
the page has mutated and is no longer scrollable when it arrives.
The next update from the UI process attempts to fix this, but since it's a
syncPosition update, it doesn't clear the existing (and invalid) scroll
position.

syncPosition GraphicsLayer updates should mutate the layer without marking it as
changed (to prevent it being sent back to the UI-process from where it
originated).  They also had the side effect of not overriding existing
approximatePosition updates, which seems to be unexpected. This makes the
behavior match setPosition, where it clears any prior approximate position
updates.

* Source/WebCore/platform/graphics/GraphicsLayer.h:
(WebCore::GraphicsLayer::syncPosition):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::syncPosition):

Canonical link: https://commits.webkit.org/302159@main
